### PR TITLE
[addons] fix event handling

### PR DESCRIPTION
**Pull request by tamland**
Should hopefully fix all the various updating issues stemming from the lack of a event strategy in the addon system.
This includes a very basic pub/sub utility for sending events; it gets the job done but if anyone know something better, existing, that can be used, free to suggest one.

### CODE REVIEW

**Comment by tamland** [comment]
can someone provide xcode/vs for this?

File: xbmc/listproviders/DirectoryProvider.cpp [hunk 1]
```diff
@@ -253,6 +260,19 @@ void CDirectoryProvider::Fetch(std::vector<CGUIListItemPtr> &items) const
   }
 }
 
+void CDirectoryProvider::OnEvent(const ADDON::AddonEvent& event)
+{
+  CSingleLock lock(m_section);
```
Comment (wsnipex) [inline xbmc/listproviders/DirectoryProvider.cpp]: couldn't the lock be moved down inside the if?

Comment (tamland) [inline xbmc/listproviders/DirectoryProvider.cpp]: m_currentUrl needs to be protected as well, since it may be written to

**Comment by ksooo** [comment]
@tamland you can cherry-pick the xcode project update for this PR here https://github.com/ksooo/xbmc/commit/1315dc065439d05f623637a5f8f31f1b852caf17 if you like.

**Comment by tamland** [comment]
thanks @ksooo!

**Comment by tamland** [comment]
jenkins build this please

**Comment by MartijnKaijser** [comment]
Did quick test and seems to work fine now with installing Addons and updating vfs

**Comment by da-anda** [comment]
sorry, haven't checked the changes - will this also update the add-on lists in terms of "recently used" on home screen? IIRC this is also not working atm (at least last time I checked).
Also, at least for me PVR channels and file shares in "videos" node not always show up, but that's a different issue.

**Comment by MartijnKaijser** [comment]
@da-anda yep for add-on question :)

**Comment by MartijnKaijser** [comment]
Can we include this for alpha3?

**Comment by tamland** [comment]
yes. just had to squash.

**Comment by BigNoid** [comment]
This PR causes some issues. Easiest to reproduce is starting a video and do a ```ReloadSkin()```. Without this PR it works fine, with this PR it crashes.

**Pull request merged**